qubes/storage/callback.py

```diff
@@ -0,0 +1,442 @@
+            self._callback('on_sinit')
+        self._cb_requires_init = False
+
+    def _assertInitialized(self, **kwargs):
+        if self._cb_requires_init:
+            self._init(**kwargs)
+
+    def _callback(self, cb, cb_args=[], log=logging.getLogger('qubes.storage.callback')):
+        '''Run a callback.
+        :param cb: Callback identifier string.
+        :param cb_args: Optional arguments to pass to the command as first arguments.
+                        Only passed on for the generic command specified as `cmd`, not for `on_xyz` callbacks.
+        '''
+        if self._cb_ctor_done:
+            cmd = self._cb_conf.get(cb)
+            args = [] #on_xyz callbacks should never receive arguments
+            if not cmd:
+                cmd = self._cb_conf.get('cmd')
+                args = [ self.name, self._cb_conf['bdriver'], cb, self._cb_cmd_arg, *cb_args ]
+            if cmd and cmd != '-':
+                args = filter(None, args)
+                args = ' '.join(quote(str(a)) for a in args)
+                cmd = ' '.join(filter(None, [cmd, args]))
+                log.info('callback driver executing (%s, %s %s): %s' % (self._cb_conf_id, cb, cb_args, cmd))
+                res = subprocess.run(cmd, shell=True, check=True, executable='/bin/bash', stdout=subprocess.PIPE, stderr=subprocess.PIPE, universal_newlines=True)
+                #stdout & stderr are reported if the exit code check fails
+                log.debug('callback driver stdout (%s, %s %s): %s' % (self._cb_conf_id, cb, cb_args, res.stdout))
+                log.debug('callback driver stderr (%s, %s %s): %s' % (self._cb_conf_id, cb, cb_args, res.stderr))
+                if self._cb_conf.get('signal_back', False) is True:
+                    self._process_signals(res.stdout, log)
+
+    def _process_signals(self, out, log=logging.getLogger('qubes.storage.callback')):
+        '''Process any signals found inside a string.
+        :param out: String to check for signals. Each signal must be on a dedicated line.
+                    They are executed in the order they are found. Callbacks are not triggered.
+        '''
+        for line in out.splitlines():
+            if line == 'SIGNAL_setup':
+                log.info('callback driver processing SIGNAL_setup for %s' % self._cb_conf_id)
+                self.setup(callback=False)
+
+    def __del__(self):
+        s = super()
+        if hasattr(s, '__del__'):
+            s.__del__()
+
+    @property
+    def config(self):
+        return {
+            'name': self.name,
+            'driver': CallbackPool.driver,
+            'conf_id': self._cb_conf_id,
+        }
+
+    def destroy(self):
+        self._assertInitialized()
+        ret = self._cb_impl.destroy()
+        self._callback('on_destroy')
+        return ret
+
+    def init_volume(self, vm, volume_config):
+        return CallbackVolume(self, self._cb_impl.init_volume(vm, volume_config))
+
+    def setup(self, callback=True):
```

Comment:
Please keep original parameters (none here). If you need to call the same from elsewhere, with custom params, add a helper function with those params, then call it here too.